import essential libraries

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
import seaborn as sb
import plotly.graph_objects as go
from sklearn import preprocessing,metrics,feature_selection
from sklearn.model_selection import train_test_split
from scipy import stats
from sklearn_pandas import CategoricalImputer

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


Read the dataset, and choose 12 random feature which are 


1.   host_id
2.   host_response_time
3. host_response_rate
4. host_acceptance_rate
5. property_type
6. room_type
7. beds
8. extra_people
9. number_of_reviews
10. number_of_reviews_ltm
11. reviews_per_month 
12. neighbourhood








In [ ]:
#get the Train dataset
data =pd.read_csv("dataset.csv",dtype=object)
temp = data[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood","price","review_scores_rating"]]
features = pd.DataFrame(temp)
print(features.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13049 entries, 0 to 13048
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   host_id                12811 non-null  object
 1   host_response_time     12199 non-null  object
 2   host_response_rate     12246 non-null  object
 3   host_acceptance_rate   12729 non-null  object
 4   property_type          13048 non-null  object
 5   room_type              13045 non-null  object
 6   beds                   13015 non-null  object
 7   extra_people           13027 non-null  object
 8   number_of_reviews      13049 non-null  object
 9   number_of_reviews_ltm  13049 non-null  object
 10  reviews_per_month      11110 non-null  object
 11  neighbourhood          11861 non-null  object
 12  price                  12847 non-null  object
 13  review_scores_rating   11075 non-null  object
dtypes: object(14)
memory usage: 1.4+ MB
None


converting all continous data feature into float64 type and convert string data into null

In [ ]:
features['host_response_rate']=features['host_response_rate'].str.replace("%","")
features['host_response_rate']=pd.to_numeric(features['host_response_rate'], errors='coerce')

features['host_acceptance_rate']=features['host_acceptance_rate'].str.replace("%","")
features['host_acceptance_rate']=pd.to_numeric(features['host_acceptance_rate'], errors='coerce')

features['beds']=pd.to_numeric(features['beds'], errors='coerce')

features['extra_people']=pd.to_numeric(features['extra_people'], errors='coerce')

features['reviews_per_month']=pd.to_numeric(features['reviews_per_month'], errors='coerce')

features['number_of_reviews']=pd.to_numeric(features['reviews_per_month'], errors='coerce')

features['number_of_reviews_ltm']=pd.to_numeric(features['reviews_per_month'], errors='coerce')


Check for number of NULL values in each feature

In [ ]:
for i in features:
  print(i,features[i].isna().sum())

host_id 238
host_response_time 850
host_response_rate 1092
host_acceptance_rate 428
property_type 1
room_type 4
beds 97
extra_people 12821
number_of_reviews 1977
number_of_reviews_ltm 1977
reviews_per_month 1977
neighbourhood 1188
price 202
review_scores_rating 1974


all collumns are required to be processed further.

we will start from the least amount of missing values which is property_type. since there is only 1 data missing which means that by removing the missing value in property type would not reduce any significant information.

In [ ]:
features.dropna(subset = ["property_type"], inplace=True)
data.dropna(subset = ["property_type"], inplace=True)
for i in features:
  print(i,features[i].isna().sum())
print(features.head())

host_id 238
host_response_time 849
host_response_rate 1091
host_acceptance_rate 427
property_type 0
room_type 4
beds 96
extra_people 12821
number_of_reviews 1976
number_of_reviews_ltm 1976
reviews_per_month 1976
neighbourhood 1188
price 202
review_scores_rating 1974
   host_id  host_response_time  ...       price  review_scores_rating
0   151977                 NaN  ...   $4,183.00                    94
1   964081  within a few hours  ...  $11,048.00                    95
2   341577  within a few hours  ...   $4,171.00                    95
3  1573631        within a day  ...   $6,952.00                    95
4   341577  within a few hours  ...   $3,957.00                    96

[5 rows x 14 columns]


next we will consider categorical data which mean we will replace the data with the most frequent data in the dataset

catergorical data are:

1. host_id
2. host response time
3. room_type
4. neighbourhood

In [ ]:
#replacing with the frequently occured
features['host_id']=features['host_id'].fillna(features['host_id'].mode()[0])
features['room_type']=features['room_type'].fillna(features['room_type'].mode()[0])
features['host_response_time']=features['host_response_time'].fillna(features['host_response_time'].mode()[0])
features['neighbourhood']=features['neighbourhood'].fillna(features['neighbourhood'].mode()[0])
for i in features:
  print(i,features[i].isna().sum())


host_id 0
host_response_time 0
host_response_rate 1091
host_acceptance_rate 427
property_type 0
room_type 0
beds 96
extra_people 12821
number_of_reviews 1976
number_of_reviews_ltm 1976
reviews_per_month 1976
neighbourhood 0
price 202
review_scores_rating 1974


next we will consider continous data and replace null values with the median

In [ ]:
features['host_response_rate'] = features['host_response_rate'].fillna(features['host_response_rate'].median())

features['host_acceptance_rate'] = features['host_acceptance_rate'].fillna(features['host_acceptance_rate'].median())

features['beds'] = features['beds'].fillna(features['beds'].median())

features['extra_people'] = features['extra_people'].fillna(features['extra_people'].median())

features['reviews_per_month'] = features['reviews_per_month'].fillna(features['reviews_per_month'].median())

features['number_of_reviews'] = features['number_of_reviews'].fillna(features['number_of_reviews'].median())

features['number_of_reviews_ltm'] = features['number_of_reviews_ltm'].fillna(features['number_of_reviews_ltm'].median())

for i in features:
  print(i,features[i].isna().sum())


host_id 0
host_response_time 0
host_response_rate 0
host_acceptance_rate 0
property_type 0
room_type 0
beds 0
extra_people 0
number_of_reviews 0
number_of_reviews_ltm 0
reviews_per_month 0
neighbourhood 0
price 202
review_scores_rating 1974


show the data after Null values are removed

In [ ]:
print(features.info())
#check for continous data
display(features.describe())


<class 'pandas.core.frame.DataFrame'>
Int64Index: 13048 entries, 0 to 13048
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   host_id                13048 non-null  object 
 1   host_response_time     13048 non-null  object 
 2   host_response_rate     13048 non-null  float64
 3   host_acceptance_rate   13048 non-null  float64
 4   property_type          13048 non-null  object 
 5   room_type              13048 non-null  object 
 6   beds                   13048 non-null  float64
 7   extra_people           13048 non-null  float64
 8   number_of_reviews      13048 non-null  float64
 9   number_of_reviews_ltm  13048 non-null  float64
 10  reviews_per_month      13048 non-null  float64
 11  neighbourhood          13048 non-null  object 
 12  price                  12846 non-null  object 
 13  review_scores_rating   11074 non-null  object 
dtypes: float64(7), object(7)
memory usage: 1.5+ MB
None


host_response_rate  ...  reviews_per_month
count        1.304800e+04  ...       13048.000000
mean         8.202861e+04  ...           1.712597
std          3.823178e+06  ...           3.106943
min          0.000000e+00  ...           0.000000
25%          1.000000e+02  ...           0.800000
50%          1.000000e+02  ...           1.510000
75%          1.000000e+02  ...           2.240000
max          1.814470e+08  ...         227.000000

[8 rows x 7 columns]

apply odrinal encoder to categorical data

In [ ]:
#apply encoder
feature_oe = preprocessing.OrdinalEncoder()
feature_oe.fit(features[['host_id','host_response_time','property_type','room_type','neighbourhood']])
features[['host_id','host_response_time','property_type','room_type','neighbourhood']] = feature_oe.transform(features[['host_id','host_response_time','property_type','room_type','neighbourhood']])
display(features.head())

host_id  host_response_time  ...       price  review_scores_rating
0    411.0                85.0  ...   $4,183.00                    94
1   3015.0                84.0  ...  $11,048.00                    95
2   2266.0                84.0  ...   $4,171.00                    95
3    458.0                83.0  ...   $6,952.00                    95
4   2266.0                84.0  ...   $3,957.00                    96

[5 rows x 14 columns]

Split the dataset between training and testing

In [ ]:
Train, Test = train_test_split(features, test_size=0.2)
Train.to_csv(r'Train.csv', index = False)
Test.to_csv(r'Test.csv', index = False)

**Feature engineering**

Get the data

In [ ]:
#get the data
Train_Raw =pd.read_csv("Train.csv")
Test_Raw=pd.read_csv("Test.csv")
display(Train_Raw.head())

host_id  host_response_time  ...       price  review_scores_rating
0   2482.0                83.0  ...   $5,041.00                    82
1   2357.0                85.0  ...   $6,631.00                    70
2   1310.0                85.0  ...   $6,952.00                    80
3   1773.0                85.0  ...  $13,049.00                   100
4   1098.0                85.0  ...  $19,951.00                    96

[5 rows x 14 columns]

from we can see above based one the mean standard deviation and maximum value. there is outliers in the continous value. therefore we need to remove those outliers

In [ ]:
temp=pd.DataFrame(Train_Raw[['host_response_rate','host_acceptance_rate','beds','extra_people','reviews_per_month','number_of_reviews','number_of_reviews_ltm']])
z_scores = stats.zscore(temp)
abs_z_scores = np.abs(z_scores)
filtered_entries = (abs_z_scores < 3).all(axis=1)
new_features = Train_Raw[filtered_entries]
display(new_features.describe())
print(new_features.info())

host_id  host_response_time  ...  reviews_per_month  neighbourhood
count  10357.000000        10357.000000  ...       10357.000000   10357.000000
mean    1527.968620           84.261852  ...           1.652083      51.852757
std      862.332903            5.987997  ...           1.139071      23.522292
min        3.000000            0.000000  ...           0.000000       1.000000
25%      781.000000           85.000000  ...           0.810000      31.000000
50%     1501.000000           85.000000  ...           1.510000      63.000000
75%     2303.000000           85.000000  ...           2.240000      73.000000
max     3099.000000          101.000000  ...           8.000000     109.000000

[8 rows x 12 columns]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10357 entries, 0 to 10437
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   host_id                10357 non-null  float64
 1   host_response_time     10357 non-null  float64
 2   host_response_rate     10357 non-null  float64
 3   host_acceptance_rate   10357 non-null  float64
 4   property_type          10357 non-null  float64
 5   room_type              10357 non-null  float64
 6   beds                   10357 non-null  float64
 7   extra_people           10357 non-null  float64
 8   number_of_reviews      10357 non-null  float64
 9   number_of_reviews_ltm  10357 non-null  float64
 10  reviews_per_month      10357 non-null  float64
 11  neighbourhood          10357 non-null  float64
 12  price                  10236 non-null  object 
 13  review_scores_rating   8784 non-null   object 
dtypes: float64(12), object(2)
memory usage: 1.2+ MB
None


we will also remove any outlier for the categorical data by removing any category which only have the frequency of one. Since the data category is too diffrent than the others

In [ ]:

#host id
# Frequencies of each value in the column.
freq = new_features['host_id'].value_counts()
# Select frequent values. Value is in the index.
frequent_values = freq[freq > 1].index
# Use only rows with value frequency above threshold.
new_features=new_features[new_features['host_id'].isin(frequent_values)]


#apply for the other categorical data
#host_response_time
freq = new_features['host_response_time'].value_counts()
frequent_values = freq[freq > 1].index
new_features=new_features[new_features['host_response_time'].isin(frequent_values)]


#property_type
freq = new_features['property_type'].value_counts()
frequent_values = freq[freq > 1].index
new_features=new_features[new_features['property_type'].isin(frequent_values)]


#room_type
freq = new_features['room_type'].value_counts()
frequent_values = freq[freq > 1].index
new_features=new_features[new_features['room_type'].isin(frequent_values)]


#neighbourhood
freq = new_features['neighbourhood'].value_counts()
frequent_values = freq[freq > 1].index
new_features=new_features[new_features['neighbourhood'].isin(frequent_values)]


print(new_features.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9064 entries, 0 to 10437
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   host_id                9064 non-null   float64
 1   host_response_time     9064 non-null   float64
 2   host_response_rate     9064 non-null   float64
 3   host_acceptance_rate   9064 non-null   float64
 4   property_type          9064 non-null   float64
 5   room_type              9064 non-null   float64
 6   beds                   9064 non-null   float64
 7   extra_people           9064 non-null   float64
 8   number_of_reviews      9064 non-null   float64
 9   number_of_reviews_ltm  9064 non-null   float64
 10  reviews_per_month      9064 non-null   float64
 11  neighbourhood          9064 non-null   float64
 12  price                  8971 non-null   object 
 13  review_scores_rating   7595 non-null   object 
dtypes: float64(12), object(2)
memory usage: 1.0+ MB
None


scale the **data**

In [ ]:


# Scaled feature 
temp_feature=Train_Raw[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
feature_min_max_scaler = preprocessing.MinMaxScaler(feature_range =(0, 1)) 
feature_min_max_scaler.fit(temp_feature) 
feature_scaled = feature_min_max_scaler.transform(temp_feature)
feature_scaled=pd.DataFrame(feature_scaled)
feature_scaled.columns=temp_feature.columns
Train=feature_scaled.join(Train_Raw[["price","review_scores_rating"]])
display(Train.head())

temp_feature=Test_Raw[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
feature_scaled = feature_min_max_scaler.transform(temp_feature)
feature_scaled=pd.DataFrame(feature_scaled)
feature_scaled.columns=temp_feature.columns
Test=feature_scaled.join(Test_Raw[["price","review_scores_rating"]])
display(Test.head())


host_id  host_response_time  ...       price  review_scores_rating
0  0.800904            0.821782  ...   $5,041.00                    82
1  0.760568            0.841584  ...   $6,631.00                    70
2  0.422717            0.841584  ...   $6,952.00                    80
3  0.572120            0.841584  ...  $13,049.00                   100
4  0.354308            0.841584  ...  $19,951.00                    96

[5 rows x 14 columns]

host_id  host_response_time  ...       price  review_scores_rating
0  0.376573            0.831683  ...   $3,957.00                    90
1  0.809939            0.841584  ...   $4,920.00                   100
2  0.408196            0.841584  ...  $49,984.00                    84
3  0.313972            0.831683  ...   $9,975.00                   NaN
4  0.655050            0.841584  ...   $8,045.00                   100

[5 rows x 14 columns]

**Logistic Regression**

convert price into float datatype 

In [ ]:
Train['price']=Train['price'].str.replace('$',"")
Train['price']=Train['price'].str.replace(',',"")
Train['price']=pd.to_numeric(Train['price'], errors='coerce')

Train_Raw['price']=Train_Raw['price'].str.replace('$',"")
Train_Raw['price']=Train_Raw['price'].str.replace(',',"")
Train_Raw['price']=pd.to_numeric(Train_Raw['price'], errors='coerce')

Test['price']=Test['price'].str.replace('$',"")
Test['price']=Test['price'].str.replace(',',"")
Test['price']=pd.to_numeric(Test['price'], errors='coerce')

Test_Raw['price']=Test_Raw['price'].str.replace('$',"")
Test_Raw['price']=Test_Raw['price'].str.replace(',',"")
Test_Raw['price']=pd.to_numeric(Test_Raw['price'], errors='coerce')

lr_Train=Train.copy()
lr_Train_Raw=Train_Raw.copy()
lr_Test=Test.copy()
lr_Test_Raw=Test_Raw.copy()
lr_Train.info()

lr_Train.head()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10438 entries, 0 to 10437
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   host_id                10438 non-null  float64
 1   host_response_time     10438 non-null  float64
 2   host_response_rate     10438 non-null  float64
 3   host_acceptance_rate   10438 non-null  float64
 4   property_type          10438 non-null  float64
 5   room_type              10438 non-null  float64
 6   beds                   10438 non-null  float64
 7   extra_people           10438 non-null  float64
 8   number_of_reviews      10438 non-null  float64
 9   number_of_reviews_ltm  10438 non-null  float64
 10  reviews_per_month      10438 non-null  float64
 11  neighbourhood          10438 non-null  float64
 12  price                  10197 non-null  float64
 13  review_scores_rating   8861 non-null   object 
dtypes: float64(13), object(1)
memory usage: 1.1+ MB


host_id  host_response_time  ...    price  review_scores_rating
0  0.800904            0.821782  ...   5041.0                    82
1  0.760568            0.841584  ...   6631.0                    70
2  0.422717            0.841584  ...   6952.0                    80
3  0.572120            0.841584  ...  13049.0                   100
4  0.354308            0.841584  ...  19951.0                    96

[5 rows x 14 columns]

Check for NULL values

In [ ]:
print(lr_Train['price'].isna().sum())

241


drop null values

In [ ]:
lr_Train.dropna(subset = ["price"], inplace=True)
lr_Train_Raw.dropna(subset = ["price"], inplace=True)
lr_Test.dropna(subset = ["price"], inplace=True)
lr_Test_Raw.dropna(subset = ["price"], inplace=True)

print(lr_Train['price'].isna().sum())


0


remove outliers

In [ ]:
z_scores = stats.zscore(lr_Train['price'])
abs_z_scores = np.abs(z_scores)
filtered_entries = (abs_z_scores < 3)
lr_feature_train = lr_Train[filtered_entries]
display(lr_feature_train.info())
display(lr_feature_train.describe())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10075 entries, 0 to 10437
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   host_id                10075 non-null  float64
 1   host_response_time     10075 non-null  float64
 2   host_response_rate     10075 non-null  float64
 3   host_acceptance_rate   10075 non-null  float64
 4   property_type          10075 non-null  float64
 5   room_type              10075 non-null  float64
 6   beds                   10075 non-null  float64
 7   extra_people           10075 non-null  float64
 8   number_of_reviews      10075 non-null  float64
 9   number_of_reviews_ltm  10075 non-null  float64
 10  reviews_per_month      10075 non-null  float64
 11  neighbourhood          10075 non-null  float64
 12  price                  10075 non-null  float64
 13  review_scores_rating   8580 non-null   object 
dtypes: float64(13), object(1)
memory usage: 1.2+ MB


None

host_id  host_response_time  ...  neighbourhood          price
count  10075.000000        10075.000000  ...   10075.000000   10075.000000
mean       0.488390            0.838923  ...       0.463087   19035.607465
std        0.278470            0.028745  ...       0.213012   24246.193675
min        0.000968            0.019802  ...       0.000000       0.000000
25%        0.249919            0.841584  ...       0.275229    6757.000000
50%        0.477573            0.841584  ...       0.568807   11048.000000
75%        0.742820            0.841584  ...       0.605505   19951.000000
max        1.000000            0.990099  ...       0.798165  250025.000000

[8 rows x 13 columns]

Feature selection for categorical features use ANOVA and for continous features use pearson.

When using ANOVA we will take the p value and the close the value get to 0 the better

When using pearson correlation we want to have the correlation coefficient to be as close to 1 as possible

In [ ]:
categorical=lr_feature_train[['host_id','host_response_time','property_type','room_type','neighbourhood']]
continous =lr_feature_train[['host_response_rate','host_acceptance_rate','beds','extra_people','reviews_per_month','number_of_reviews','number_of_reviews_ltm']]
label=lr_feature_train[["price"]]


fval,pval=feature_selection.f_classif(categorical,label.values.ravel())
pvalue=pd.DataFrame(categorical.columns,columns=['feature'])
pval=pd.DataFrame(pval,columns=['p value'])
display(pvalue.join(pval))
display(pd.DataFrame(continous.join(label).corr()['price']))




feature        p value
0             host_id   1.009943e-77
1  host_response_time   0.000000e+00
2       property_type  5.312683e-152
3           room_type   0.000000e+00
4       neighbourhood   6.794549e-39

price
host_response_rate    -0.017473
host_acceptance_rate   0.079334
beds                   0.085022
extra_people           0.013014
reviews_per_month      0.039731
number_of_reviews      0.039731
number_of_reviews_ltm  0.039731
price                  1.000000

In [ ]:
selected_feature=lr_feature_train[["host_id","host_response_time","host_acceptance_rate","property_type","room_type","beds", "neighbourhood"]]

from the above value we will take all the categorical data since all of the values are very close to 0.

for the continous we will only take the correlation above 0.05 which are host acceptance rate and beds

Train logisitic regression with feature engineered data

In [ ]:
display(selected_feature.describe())
display(lr_Test.describe())

host_id  host_response_time  ...          beds  neighbourhood
count  10075.000000        10075.000000  ...  1.007500e+04   10075.000000
mean       0.488390            0.838923  ...  2.426709e-06       0.463087
std        0.278470            0.028745  ...  3.481833e-06       0.213012
min        0.000968            0.019802  ...  0.000000e+00       0.000000
25%        0.249919            0.841584  ...  8.620593e-07       0.275229
50%        0.477573            0.841584  ...  1.724119e-06       0.568807
75%        0.742820            0.841584  ...  2.586178e-06       0.605505
max        1.000000            0.990099  ...  1.205896e-04       0.798165

[8 rows x 7 columns]

host_id  host_response_time  ...  neighbourhood          price
count  2554.000000         2554.000000  ...    2554.000000    2554.000000
mean      0.486886            0.839107  ...       0.463500   27272.433634
std       0.280105            0.026046  ...       0.215669   83927.654639
min       0.001936            0.158416  ...       0.000000       0.000000
25%       0.244918            0.841584  ...       0.275229    6952.000000
50%       0.469184            0.841584  ...       0.568807   11048.000000
75%       0.738061            0.841584  ...       0.660550   19951.000000
max       0.984511            0.990099  ...       0.807339  999995.000000

[8 rows x 13 columns]

In [ ]:
XTrain=selected_feature
YTrain = lr_feature_train[["price"]]
XTest=lr_Test[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
YTest = lr_Test[["price"]]

In [ ]:
from sklearn.linear_model import LogisticRegression
LGR = LogisticRegression(max_iter=1000).fit(XTrain, YTrain.astype('int').values.ravel())

In [ ]:
CTest  = LGR.predict(XTest[["host_id","host_response_time","host_acceptance_rate","property_type","room_type","beds", "neighbourhood"]])
CTrain  = LGR.predict(XTrain[["host_id","host_response_time","host_acceptance_rate","property_type","room_type","beds", "neighbourhood"]])

print("Train set acc: ", metrics.accuracy_score(YTrain.astype('int'), CTrain))
print("Test set acc: ", metrics.accuracy_score(YTest.astype('int'), CTest))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(YTest, CTest)))


Train set acc:  0.05349875930521092
Test set acc:  0.056382145653876274
Root Mean Squared Error: 85191.91037470057


Train logisitic regression with raw data

In [ ]:
display(lr_Train_Raw.head())
display(lr_Test_Raw.head())

host_id  host_response_time  ...    price  review_scores_rating
0   2482.0                83.0  ...   5041.0                    82
1   2357.0                85.0  ...   6631.0                    70
2   1310.0                85.0  ...   6952.0                    80
3   1773.0                85.0  ...  13049.0                   100
4   1098.0                85.0  ...  19951.0                    96

[5 rows x 14 columns]

host_id  host_response_time  ...    price  review_scores_rating
0   1167.0                84.0  ...   3957.0                    90
1   2510.0                85.0  ...   4920.0                   100
2   1265.0                85.0  ...  49984.0                    84
3    973.0                84.0  ...   9975.0                   NaN
4   2030.0                85.0  ...   8045.0                   100

[5 rows x 14 columns]

In [ ]:
XTrain=lr_Train_Raw[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
YTrain = lr_Train_Raw[["price"]]
XTest=lr_Test_Raw[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
YTest = lr_Test_Raw[["price"]]

In [ ]:
from sklearn.linear_model import LogisticRegression
LGR = LogisticRegression(max_iter=1000).fit(XTrain, YTrain.astype('int').values.ravel())

In [ ]:
CTest  = LGR.predict(XTest)
CTrain  = LGR.predict(XTrain)

print("Train set acc: ", metrics.accuracy_score(YTrain.astype('int'), CTrain))
print("Test set acc: ", metrics.accuracy_score(YTest.astype('int'), CTest))
print('Root Mean Squared Error:', np.sqrt(metrics.mean_squared_error(YTest, CTest)))


Train set acc:  0.00019613611846621556
Test set acc:  0.0
Root Mean Squared Error: 88227.04403958295


Based on the result above we can see the diffrence between using raw data and data that has been proccessed. we will first be using confusion matrix as a metric to understand how much data is classified correctly since we are using logistic regression, and logistic regression is a classifier model. After that we will use Root mean square error to understand how effective it is in predicting continous data.

the result for using rawdata  shows a horrible results as almost all of them are not classified properly. On the other hand after featured engineered the classification accuracy increase up to 56% for using test data and 53% for using training data. However the root mean square error does not show much diffrence as approximately 88227 for raw data and 85191 for processed data. there is an increase in performance by a reduction of approximately 3000 in the rmse.

**Get review score rating data for question no 3,4,5**

convert data type to float

In [ ]:
Train['review_scores_rating']=pd.to_numeric(Train['review_scores_rating'], errors='coerce')
Test['review_scores_rating']=pd.to_numeric(Test['review_scores_rating'], errors='coerce')
Train_Raw['review_scores_rating']=pd.to_numeric(Train['review_scores_rating'], errors='coerce')
Test_Raw['review_scores_rating']=pd.to_numeric(Test['review_scores_rating'], errors='coerce')
clf_Train=Train.copy()
clf_Train_Raw=Train_Raw.copy()
clf_Test=Test.copy()
clf_Test_Raw=Test_Raw.copy()

clf_Train.describe()

host_id  host_response_time  ...          price  review_scores_rating
count  10438.000000        10438.000000  ...   10197.000000           8670.000000
mean       0.493226            0.833649  ...   26600.909798             93.547866
std        0.277898            0.063410  ...   79521.022528             28.196957
min        0.000000            0.000000  ...       0.000000              0.000000
25%        0.252339            0.841584  ...    6865.000000             90.000000
50%        0.484350            0.841584  ...   11551.000000             95.000000
75%        0.742820            0.841584  ...   20001.000000             98.000000
max        1.000000            1.000000  ...  999995.000000           1125.000000

[8 rows x 14 columns]

check NULL values and drop them

In [ ]:
#print(clf_Train['review_scores_rating'].isna().sum())
clf_Train.dropna(subset = ["review_scores_rating"], inplace=True)
clf_Test.dropna(subset = ["review_scores_rating"], inplace=True)
clf_Train_Raw.dropna(subset = ["review_scores_rating"], inplace=True)
clf_Test_Raw.dropna(subset = ["review_scores_rating"], inplace=True)

clf_Train.info()
#print(clf_Train['review_scores_rating'].isna().sum())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8670 entries, 0 to 10437
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   host_id                8670 non-null   float64
 1   host_response_time     8670 non-null   float64
 2   host_response_rate     8670 non-null   float64
 3   host_acceptance_rate   8670 non-null   float64
 4   property_type          8670 non-null   float64
 5   room_type              8670 non-null   float64
 6   beds                   8670 non-null   float64
 7   extra_people           8670 non-null   float64
 8   number_of_reviews      8670 non-null   float64
 9   number_of_reviews_ltm  8670 non-null   float64
 10  reviews_per_month      8670 non-null   float64
 11  neighbourhood          8670 non-null   float64
 12  price                  8646 non-null   float64
 13  review_scores_rating   8670 non-null   float64
dtypes: float64(14)
memory usage: 1016.0 KB


check and drop outliers.
Since it is a rating we can assume that the maximum value should be 100 as in 100% and the minimum as 0. so any value above 100 should be removed

In [ ]:
clf_feature_train = clf_Train[clf_Train['review_scores_rating']<=100]
clf_feature_train = clf_feature_train[clf_Train['review_scores_rating']>=0]
display(clf_feature_train.describe())
display(clf_feature_train.info())

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: UserWarning:

Boolean Series key will be reindexed to match DataFrame index.



host_id  host_response_time  ...          price  review_scores_rating
count  8649.000000         8649.000000  ...    8628.000000           8649.000000
mean      0.481851            0.838765  ...   24882.839534             92.543762
std       0.283739            0.032903  ...   75288.039578              8.931863
min       0.001936            0.039604  ...       0.000000              0.000000
25%       0.237496            0.841584  ...    6822.250000             90.000000
50%       0.460148            0.841584  ...   11048.000000             95.000000
75%       0.757341            0.841584  ...   19951.000000             98.000000
max       0.986770            0.990099  ...  999995.000000            100.000000

[8 rows x 14 columns]

<class 'pandas.core.frame.DataFrame'>
Int64Index: 8649 entries, 0 to 10437
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   host_id                8649 non-null   float64
 1   host_response_time     8649 non-null   float64
 2   host_response_rate     8649 non-null   float64
 3   host_acceptance_rate   8649 non-null   float64
 4   property_type          8649 non-null   float64
 5   room_type              8649 non-null   float64
 6   beds                   8649 non-null   float64
 7   extra_people           8649 non-null   float64
 8   number_of_reviews      8649 non-null   float64
 9   number_of_reviews_ltm  8649 non-null   float64
 10  reviews_per_month      8649 non-null   float64
 11  neighbourhood          8649 non-null   float64
 12  price                  8628 non-null   float64
 13  review_scores_rating   8649 non-null   float64
dtypes: float64(14)
memory usage: 1013.6 KB


None

since we will be using classification algorithm. it's best to convert the output data into categorical data. we can do this by applying bins which allows us to convert the data into 5 category which are '1 star','2 star','3 star','4 star',and '5 star'. with an interval approximately of 20

In [ ]:
cut_bins=[-9999999,20,40,60,80,9999999]
cut_labels=['1 star','2 star','3 star','4 star', '5 star']
le = preprocessing.LabelEncoder()

rating_category=pd.cut(clf_feature_train['review_scores_rating'],bins=cut_bins,labels=cut_labels)
clf_feature_train['review_scores_rating']=le.fit_transform(rating_category)


rating_category=pd.cut(clf_Test['review_scores_rating'],bins=cut_bins,labels=cut_labels)
clf_Test['review_scores_rating']=le.transform(rating_category)


rating_category=pd.cut(clf_Train_Raw['review_scores_rating'],bins=cut_bins,labels=cut_labels)
clf_Train_Raw['review_scores_rating']=le.transform(rating_category)

rating_category=pd.cut(clf_Test_Raw['review_scores_rating'],bins=cut_bins,labels=cut_labels)
clf_Test_Raw['review_scores_rating']=le.transform(rating_category)




Feature selection using ANOVA for continous feature and chisquare for categorical feature

In [ ]:
categorical=clf_feature_train[['host_id','host_response_time','property_type','room_type','neighbourhood']]
continous =clf_feature_train[['host_response_rate','host_acceptance_rate','beds','extra_people','reviews_per_month','number_of_reviews','number_of_reviews_ltm']]
label=clf_feature_train[["review_scores_rating"]]


fval,pval=feature_selection.f_classif(continous,label.values.ravel())
pvalue=pd.DataFrame(categorical.columns,columns=['feature'])
pval=pd.DataFrame(pval,columns=['p value'])
display(pvalue.join(pval))

chi,pval=feature_selection.chi2(categorical,label.values.ravel())
pvalue=pd.DataFrame(categorical.columns,columns=['feature'])
pval=pd.DataFrame(pval,columns=['p value'])
display(pvalue.join(pval))

feature        p value
0             host_id   1.440618e-12
1  host_response_time  1.782760e-218
2       property_type   3.574636e-69
3           room_type   3.831613e-69
4       neighbourhood   2.365708e-86

feature   p value
0             host_id  0.841600
1  host_response_time  0.399627
2       property_type  0.974214
3           room_type  0.613478
4       neighbourhood  0.269848

all of the data shows the p value close to 0 which means it is possible to use all the features for training

split data between test and train

In [ ]:
XTrain=clf_feature_train[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
YTrain = clf_feature_train[["review_scores_rating"]]
XTest=clf_Test[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
YTest = clf_Test[["review_scores_rating"]]
display(XTrain)
display(YTrain)

host_id  host_response_time  ...  reviews_per_month  neighbourhood
0      0.800904            0.821782  ...           0.009383       0.486239
1      0.760568            0.841584  ...           0.001366       0.275229
2      0.422717            0.841584  ...           0.005419       0.733945
3      0.572120            0.841584  ...           0.010000       0.568807
4      0.354308            0.841584  ...           0.009251       0.091743
...         ...                 ...  ...                ...            ...
10428  0.801226            0.841584  ...           0.015066       0.110092
10432  0.251694            0.841584  ...           0.010705       0.055046
10433  0.749597            0.841584  ...           0.018370       0.568807
10434  0.877057            0.841584  ...           0.012159       0.688073
10437  0.344305            0.841584  ...           0.009471       0.568807

[8649 rows x 12 columns]

review_scores_rating
0                         4
1                         3
2                         3
3                         4
4                         4
...                     ...
10428                     4
10432                     4
10433                     4
10434                     4
10437                     4

[8649 rows x 1 columns]

In [ ]:
XTrain_Raw=clf_Train_Raw[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
YTrain_Raw = clf_Train_Raw[["review_scores_rating"]]
XTest_Raw=clf_Test_Raw[["host_id","host_response_time","host_response_rate","host_acceptance_rate","property_type","room_type","beds", "extra_people","number_of_reviews","number_of_reviews_ltm","reviews_per_month", "neighbourhood"]]
YTest_Raw = clf_Test_Raw[["review_scores_rating"]]
display(XTrain_Raw)
display(YTrain_Raw)

host_id  host_response_time  ...  reviews_per_month  neighbourhood
0       2482.0                83.0  ...               2.13           54.0
1       2357.0                85.0  ...               0.31           31.0
2       1310.0                85.0  ...               1.23           81.0
3       1773.0                85.0  ...               2.27           63.0
4       1098.0                85.0  ...               2.10           11.0
...        ...                 ...  ...                ...            ...
10430      3.0                85.0  ...               1.51            2.0
10432    780.0                85.0  ...               2.43            7.0
10433   2323.0                85.0  ...               4.17           63.0
10434   2718.0                85.0  ...               2.76           76.0
10437   1067.0                85.0  ...               2.15           63.0

[8670 rows x 12 columns]

review_scores_rating
0                         4
1                         3
2                         3
3                         4
4                         4
...                     ...
10430                     4
10432                     4
10433                     4
10434                     4
10437                     4

[8670 rows x 1 columns]

**KNN**

Train with feature engineered data

In [ ]:
#KNN ALGORITHM
from sklearn.neighbors import KNeighborsClassifier
Ks = 30
accTrain = np.zeros((Ks-1))
accTest = np.zeros((Ks-1))
ConfustionMx = [];
for n in range(1, Ks):    
    KNN = KNeighborsClassifier(n_neighbors = n).fit(XTrain, YTrain.astype(int).values.ravel())
    CTest  = KNN.predict(XTest)
    CTrain  = KNN.predict(XTrain)
    accTrain[n - 1] = metrics.accuracy_score(YTrain.astype(int), CTrain)
    accTest[n - 1] = metrics.accuracy_score(YTest.astype(int), CTest)
    
print("Best  ACC Training : %.2f" % accTrain.max(), ", with k = ", accTrain.argmax() + 1)
print("Best  ACC Testing : %.2f" % accTest.max(), ", with k = ", accTest.argmax() + 1)

Best  ACC Training : 1.00 , with k =  1
Best  ACC Testing : 0.92 , with k =  23


Train with raw data

In [ ]:
#KNN ALGORITHM
from sklearn.neighbors import KNeighborsClassifier
Ks = 30
accTrain = np.zeros((Ks-1))
accTest = np.zeros((Ks-1))
ConfustionMx = [];
for n in range(1, Ks):    
    KNN = KNeighborsClassifier(n_neighbors = n).fit(XTrain_Raw, YTrain_Raw.astype(int).values.ravel())
    CTest  = KNN.predict(XTest_Raw)
    CTrain  = KNN.predict(XTrain_Raw)
    accTrain[n - 1] = metrics.accuracy_score(YTrain_Raw.astype(int), CTrain)
    accTest[n - 1] = metrics.accuracy_score(YTest_Raw.astype(int), CTest)
    
print("Best  ACC Training : %.2f" % accTrain.max(), ", with k = ", accTrain.argmax() + 1)
print("Best  ACC Testing : %.2f" % accTest.max(), ", with k = ", accTest.argmax() + 1)

Best  ACC Training : 1.00 , with k =  1
Best  ACC Testing : 0.91 , with k =  21


from the result above we will be using confusion matrix to understand the accuracy of the model. from the result above we can see that there is an increase in accuracy by 0.01 by processing the data. when using the test dataset to try to predict the result. with a k of 23 for engineereed data and k=21 for raw data. this shows that using raw data is worst than featured engineered data

**SVM**

Train with feature engineered data

In [ ]:
#SVM ALGORITHM
from sklearn.svm import SVC
SVM = SVC().fit(XTrain, YTrain.astype(int).values.ravel())
CTest  = SVM.predict(XTest)
CTrain  = SVM.predict(XTrain)
print("Train set acc: ", metrics.accuracy_score(YTrain.astype(int), CTrain))
print("Test set acc: ", metrics.accuracy_score(YTest.astype(int), CTest))

Train set acc:  0.91941264886114
Test set acc:  0.9134438305709024


Train with raw data

In [ ]:
#SVM ALGORITHM
from sklearn.svm import SVC
SVM = SVC().fit(XTrain_Raw, YTrain_Raw.astype(int).values.ravel())
CTest  = SVM.predict(XTest_Raw)
CTrain  = SVM.predict(XTrain_Raw)
print("Train set acc: ", metrics.accuracy_score(YTrain_Raw.astype(int), CTrain))
print("Test set acc: ", metrics.accuracy_score(YTest_Raw.astype(int), CTest))

Train set acc:  0.9186851211072664
Test set acc:  0.9111418047882136


from the result above we will be using confusion matrix to understand the accuracy of the model. from the result above we can see that there is an increase in accuracy by 0.02 by processing the data. when using the test dataset to try to predict the result. this shows that using raw data is worst than featured engineered data

**Classification Tree**

Train with feature engineered data

In [ ]:
#Tree algorithm
from sklearn.tree import DecisionTreeClassifier
DST = DecisionTreeClassifier(criterion="gini").fit(XTrain, YTrain.astype(int).values.ravel())
CTest  = DST.predict(XTest)
CTrain  = DST.predict(XTrain)
print("Train set acc: ", metrics.accuracy_score(YTrain.astype(int), CTrain))
print("Test set acc: ", metrics.accuracy_score(YTest.astype(int), CTest))

Train set acc:  0.9971094924268702
Test set acc:  0.8476058931860037


Train with raw data

In [ ]:
#Tree algorithm
from sklearn.tree import DecisionTreeClassifier
DST = DecisionTreeClassifier(criterion="gini").fit(XTrain_Raw, YTrain_Raw.astype(int).values.ravel())
CTest  = DST.predict(XTest_Raw)
CTrain  = DST.predict(XTrain_Raw)
print("Train set acc: ", metrics.accuracy_score(YTrain_Raw.astype(int), CTrain))
print("Test set acc: ", metrics.accuracy_score(YTest_Raw.astype(int), CTest))

Train set acc:  0.9973471741637832
Test set acc:  0.8531307550644567


from the result above we will be using confusion matrix to understand the accuracy of the model. from the result above we can see that there is an decrease in accuracy by approximately 0.01 by processing the data. When using the test dataset to try to predict the result. this shows that using raw data is better than featured engineered data

**GMM**

In [ ]:
from sklearn.mixture import GaussianMixture
from sklearn.metrics import silhouette_score
Ks = 30
acc = np.zeros((Ks-1))

ConfustionMx = [];
for n in range(2, Ks):   
    GMM = GaussianMixture(n_components=n).fit(XTrain) 
    YPredict = GMM.predict(XTrain)
    acc[n - 1] = silhouette_score(XTrain, YPredict)
print("Best silhouette Training : %.2f" % acc.max(), ", with k = ", acc.argmax() + 1)

Best silhouette Training : 0.61 , with k =  2


the result shows that the training data is distributed by 2 cluster with a silhouette score of 0.61 which is decent

**K Means**

In [ ]:
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score
Ks = 30
acc = np.zeros((Ks-1))

ConfustionMx = [];
for n in range(2, Ks):   
    KMN = KMeans(n_clusters=n).fit(XTrain) 
    YPredict = KMN.predict(XTrain)
    acc[n - 1] = silhouette_score(XTrain, YPredict)
print("Best  silhouette Training : %.2f" % acc.max(), ", with k = ", acc.argmax() + 1)

Best  silhouette Training : 0.39 , with k =  3


the result shows that the training data is distributed by 3 cluster with a silhouette score of 0.39 